File: METODOS DE ORDENAÇÃO/testar_programa.py
import subprocess
from pathlib import Path


BASE = Path(__file__).resolve().parent
EXE = BASE / "departamentos.exe"
ENTRADA = "entrada_departamentos.txt"
DADOS = BASE / ENTRADA


def sem_acentos(texto):
    trocas = str.maketrans(
        "ÁÀÂÃÄÅÉÈÊËÍÌÎÏÓÒÔÕÖÚÙÛÜÇáàâãäåéèêëíìîïóòôõöúùûüç",
        "AAAAAAEEEEIIIIOOOOOUUUUCaaaaaaeeeeiiiiooooouuuuc",
    )
    return texto.translate(trocas)


def executar(entrada):
    processo = subprocess.run(
        [str(EXE)],
        input=entrada.encode("utf-8"),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        cwd=BASE,
        timeout=20,
    )
    return processo.returncode, processo.stdout.decode("utf-8", errors="replace"), processo.stderr.decode("utf-8", errors="replace")


def falhar(mensagem, saida="", erro=""):
    print(f"FALHOU: {mensagem}")
    if saida:
        print("--- STDOUT ---")
        print(saida)
    if erro:
        print("--- STDERR ---")
        print(erro)
    raise SystemExit(1)


linhas = [linha.strip().split() for linha in DADOS.read_text(encoding="utf-8").splitlines() if linha.strip()]

total = 0

for partes in linhas:
    codigo, departamento, *palavras = partes

    entrada = f"{ENTRADA}\n2\n{departamento}\n0\n"
    retorno, saida, erro = executar(entrada)
    total += 1
    if retorno != 0 or f"Codigo identificador: {codigo}" not in saida:
        falhar(f"departamento {departamento} deveria retornar codigo {codigo}", saida, erro)

    entrada_sem_acento = f"{ENTRADA}\n2\n{sem_acentos(departamento)}\n0\n"
    retorno, saida, erro = executar(entrada_sem_acento)
    total += 1
    if retorno != 0 or f"Codigo identificador: {codigo}" not in saida:
        falhar(f"departamento sem acento {departamento} deveria retornar codigo {codigo}", saida, erro)

    for palavra in palavras:
        entrada = f"{ENTRADA}\n1\n{palavra}\n0\n"
        retorno, saida, erro = executar(entrada)
        total += 1
        if retorno != 0 or f"Departamento: {departamento}" not in saida:
            falhar(f"palavra-chave {palavra} deveria retornar departamento {departamento}", saida, erro)

        entrada_sem_acento = f"{ENTRADA}\n1\n{sem_acentos(palavra)}\n0\n"
        retorno, saida, erro = executar(entrada_sem_acento)
        total += 1
        if retorno != 0 or f"Departamento: {departamento}" not in saida:
            falhar(f"palavra-chave sem acento {palavra} deveria retornar departamento {departamento}", saida, erro)

retorno, saida, erro = executar(f"{ENTRADA}\n1\nNAOEXISTE\n2\nRH\n9\n0\n")
total += 1
if retorno != 0 or "Palavra-chave nao encontrada." not in saida or "Departamento nao encontrado." not in saida or "Opcao invalida." not in saida:
    falhar("casos inexistentes e opcao invalida deveriam ser tratados", saida, erro)

retorno, saida, erro = executar("arquivo_inexistente.txt\n")
total += 1
if retorno == 0 or "Erro: Nao foi possivel abrir o arquivo informado." not in saida:
    falhar("arquivo inexistente deveria retornar erro controlado", saida, erro)

print(f"Todos os {total} testes automaticos passaram.")

File: METODOS DE ORDENAÇÃO/testar_programa.cpp
#define TESTE_AUTOMATICO
#include "departamentos.cpp"

#include <cstdlib>
#include <iostream>
#include <stdexcept>

using namespace std;

void exigir(bool condicao, const string& mensagem) {
    if (!condicao) {
        throw runtime_error(mensagem);
    }
}

void testarArquivoTexto() {
    vector<Departamento> departamentos = carregarDepartamentos("entrada_departamentos.txt");

    exigir(departamentos.size() == 10, "O arquivo TXT deveria carregar 10 departamentos.");
    exigir(departamentos[0].codigo == "001", "O primeiro codigo deveria ser 001.");
    exigir(departamentos[0].nome == "FINANCEIRO", "O primeiro departamento deveria ser FINANCEIRO.");
    exigir(departamentos[0].palavrasChave[4] == "LUCRO", "A quinta palavra de FINANCEIRO deveria ser LUCRO.");

    for (const Departamento& departamento : departamentos) {
        const Departamento* encontradoPorNome = buscarPorDepartamento(departamentos, departamento.nome);
        exigir(encontradoPorNome != nullptr, "Departamento existente nao encontrado: " + departamento.nome);
        exigir(encontradoPorNome->codigo == departamento.codigo, "Codigo incorreto para: " + departamento.nome);

        for (const string& palavra : departamento.palavrasChave) {
            const Departamento* encontradoPorPalavra = buscarPorPalavraChave(departamentos, palavra);
            exigir(encontradoPorPalavra != nullptr, "Palavra-chave existente nao encontrada: " + palavra);
            exigir(encontradoPorPalavra->nome == departamento.nome, "Departamento incorreto para palavra: " + palavra);
        }
    }

    exigir(buscarPorDepartamento(departamentos, "RH") == nullptr, "Departamento inexistente nao deveria ser encontrado.");
    exigir(buscarPorPalavraChave(departamentos, "NAOEXISTE") == nullptr, "Palavra inexistente nao deveria ser encontrada.");

    exigir(buscarPorDepartamento(departamentos, "financeiro") != nullptr, "Pesquisa deveria ignorar maiusculas/minusculas.");
    exigir(buscarPorPalavraChave(departamentos, "lucro") != nullptr, "Palavra-chave deveria ignorar maiusculas/minusculas.");
    exigir(buscarPorDepartamento(departamentos, "JURIDICO") != nullptr, "Pesquisa sem acento deveria encontrar JURIDICO.");
    exigir(buscarPorPalavraChave(departamentos, "SAIDA") != nullptr, "Palavra-chave sem acento deveria encontrar SAIDA.");
}

void testarArquivoWord() {
    vector<Departamento> departamentos = carregarDepartamentos("entrada_departamentos.docx");
    exigir(departamentos.size() == 10, "O arquivo DOCX deveria carregar 10 departamentos.");
    exigir(buscarPorDepartamento(departamentos, "FINANCEIRO") != nullptr, "DOCX deveria encontrar FINANCEIRO.");
    exigir(buscarPorPalavraChave(departamentos, "LUCRO") != nullptr, "DOCX deveria encontrar LUCRO.");
}

void testarArquivoInexistente() {
    bool lancouErro = false;

    try {
        carregarDepartamentos("arquivo_inexistente.txt");
    } catch (const exception&) {
        lancouErro = true;
    }

    exigir(lancouErro, "Arquivo inexistente deveria gerar erro controlado.");
}

int main() {
    try {
        testarArquivoTexto();
        testarArquivoWord();
        testarArquivoInexistente();
    } catch (const exception& erro) {
        cerr << "FALHOU: " << erro.what() << "\n";
        return EXIT_FAILURE;
    }

    cout << "Todos os testes automaticos em C++ passaram.\n";
    return EXIT_SUCCESS;
}
